IW3: dumped xmodel "void" cannot be linked again

When I try to create a new ff archive from the uploaded data, I get
<details>
  <summary>Log</summary>

```
[INFO] Bin folder: "D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64"
[INFO] Project folder: "D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx"
[INFO] Starting Linker for project: code_post_gfx
Adding asset search path: D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx
Adding gdt search path (Not found): D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx\source_data
Adding source search path: D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx\zone_source
Loaded zone "code_post_gfx"
Adding gdt search path (Not found): D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx\zone_raw\code_post_gfx\source_data
Reading zone definition file: code_post_gfx.zone
Processing zone definition took 26ms
Loaded localize "code_post_gfx"
Loaded vertexshader "vertcol_simple.hlsl"
Loaded pixelshader "vertcol_simple.hlsl"
Loaded vertexdecl "ppcc0t0t0"
Loaded technique "vertcol_simple2d"
Loaded techniqueset "sm2/2d"
Loaded techniqueset "2d"
Loaded image "3_cursor3"
Loaded material "ui_cursor"
Loaded image "$black_3d"
Loaded image "$black_cube"
Loaded image "$white"
Loaded image "$black"
Loaded image "$gray"
Loaded image "$identitynormalmap"
Loaded image "$pixelcostcolorcode"
Loaded rawfile "rumble/damage_heavy"
Loaded rawfile "rumble/damage_light"
Loaded rawfile "rumble/defaultweapon_fire"
Loaded rawfile "rumble/defaultweapon_melee"
Loaded rawfile "rumble/damage_heavy_h.rmb"
Loaded rawfile "rumble/damage_heavy_l.rmb"
Loaded rawfile "rumble/damage_light_h.rmb"
Loaded rawfile "rumble/damage_light_l.rmb"
Loaded rawfile "rumble/defaultweapon_fire_h.rmb"
Loaded rawfile "rumble/defaultweapon_fire_l.rmb"
Loaded rawfile "rumble/defaultweapon_melee_h.rmb"
Loaded rawfile "rumble/defaultweapon_melee_l.rmb"
Loaded rawfile "rumble/tank_rumble"
Loaded rawfile "rumble/tank_rumble_h.rmb"
Loaded rawfile "rumble/tank_rumble_l.rmb"
Loaded rawfile "rumble/mig_rumble"
Loaded rawfile "rumble/artillery_rumble"
Loaded rawfile "rumble/artillery_rumble_h.rmb"
Loaded rawfile "rumble/artillery_rumble_l.rmb"
Loaded rawfile "rumble/grenade_rumble"
Loaded rawfile "rumble/grenade_rumble_h.rmb"
Loaded rawfile "rumble/grenade_rumble_l.rmb"
Loaded rawfile "scriptdebugger/help.txt"
Loaded rawfile "shock/default.shock"
Loaded rawfile "shock/hold_breath.shock"
Loaded rawfile "shock/pain.shock"
Loaded rawfile "shock/victoryscreen.shock"
Loaded rawfile "shock/flashbang.shock"
Loaded rawfile "shock/dog_bite.shock"
Loaded rawfile "shock/chaplincheat.shock"
Loaded rawfile "radiant/keys.txt"
Loaded rawfile "codescripts/character.gsc"
Loaded rawfile "codescripts/delete.gsc"
Loaded rawfile "codescripts/struct.gsc"
Loaded rawfile "language.cfg"
Loaded rawfile "thumbstick_default.cfg"
Loaded rawfile "devgui_scriptart.cfg"
Loaded rawfile "devgui_maps_sp.cfg"
Loaded rawfile "devgui_main.cfg"
Loaded rawfile "devgui_visibility.cfg"
Loaded rawfile "devgui_renderer.cfg"
Loaded rawfile "viewsensitivity_med.cfg"
Loaded rawfile "default.cfg"
Loaded rawfile "avatar_dev.cfg"
Loaded rawfile "createfx.cfg"
Loaded rawfile "autoexec_dev.cfg"
Loaded rawfile "developer.cfg"
Loaded rawfile "thumbstick_southpaw.cfg"
Loaded rawfile "thumbstick_legacy.cfg"
Loaded rawfile "thumbstick_legacysouthpaw.cfg"
Loaded rawfile "viewsensitivity_low.cfg"
Loaded rawfile "viewsensitivity_high.cfg"
Loaded rawfile "viewsensitivity_ultra.cfg"
Loaded rawfile "devgui_aimassist.cfg"
Loaded rawfile "devgui_modelpreviewer.cfg"
Loaded rawfile "default_splitscreen.cfg"
Loaded rawfile "default_systemlink.cfg"
Loaded rawfile "default_xboxlive.cfg"
Loaded rawfile "default_480p.cfg"
Loaded rawfile "default_720p.cfg"
Loaded rawfile "ac130.cfg"
Loaded rawfile "robotg.cfg"
Loaded rawfile "chad.cfg"
Loaded rawfile "jake.cfg"
Loaded rawfile "massey.cfg"
Loaded rawfile "roger.cfg"
Loaded rawfile "jiesang.cfg"
Loaded rawfile "test.cfg"
Loaded rawfile "minimap.cfg"
Loaded rawfile "ragdoll.cfg"
Loaded rawfile "demo.cfg"
Loaded rawfile "antfarmcodmw1.cfg"
Loaded rawfile "antfarmcodmw2.cfg"
Loaded rawfile "enable_art.cfg"
Loaded soundcurve "default"
Loaded loadedsound "misc/coup_assassination_fire_00d.wav"
Loaded soundcurve ""
Loaded sound "enter_prestige"
Loaded loadedsound "misc/mouse_ylover.wav"
Loaded sound "match_countdown_tick"
Loaded loadedsound "user_interface/ui_select_v1.wav"
Loaded sound "mouse_click"
Loaded loadedsound "user_interface/ui_over_v1.wav"
Loaded sound "mouse_over"
Loaded loadedsound "user_interface/ui_sub_over_v1.wav"
Loaded sound "mouse_submenu_over"
Loaded loadedsound "mp/mp_experiencefill_onesecond.wav"
Loaded sound "mp_experience_tick"
Loaded loadedsound "mp/mp_player_join_gate.wav"
Loaded sound "mp_player_join"
Loaded loadedsound "mp/mp_player_leave_elec.wav"
Loaded sound "mp_player_leave"
Loaded sound "music_mainmenu"
Loaded sound "music_mainmenu_mp"
Loaded loadedsound "null.wav"
Loaded sound "null"
Loaded image "devfonts"
Loaded material "fonts/devfonts"
Loaded material "fonts/devfonts_glow"
Loaded font "fonts/smallDevFont"
Loaded image "gamefonts_pc"
Loaded material "fonts/gamefonts_pc"
Loaded material "fonts/gamefonts_pc_glow"
Loaded font "fonts/consoleFont"
Loaded font "fonts/bigDevFont"
Loaded font "fonts/bigFont"
Loaded font "fonts/smallFont"
Loaded font "fonts/boldFont"
Loaded font "fonts/normalFont"
Loaded font "fonts/extraBigFont"
Loaded font "fonts/objectiveFont"
Loaded vertexshader "vertcol_simple_dtex.hlsl"
Loaded technique "vertcol_simple_dtex"
Loaded vertexshader "vertcol_untextured_dtex.hlsl"
Loaded pixelshader "vertcol_untextured.hlsl"
Loaded vertexdecl "ppcc0"
Loaded technique "wireframe_solid_dtex"
Loaded techniqueset "sm2/effect_nofog"
Loaded techniqueset "effect_nofog"
Loaded image "missing_fx"
Loaded material "gfx_missing_fx"
Loaded fx "misc/missing_fx"
Loaded impactfx "default"
Loaded image "falloff_linear"
Loaded lightdef "light_dynamic"
Loaded lightdef "light_point_linear"
Loaded image "default"
Loaded material "$default"
Loaded material "white"
Loaded material "black"
Loaded vertexshader "vertcol_simple_tools.hlsl"
Loaded pixelshader "vertcol_shaded.hlsl"
Loaded technique "vertcol_shaded_tools"
Loaded vertexshader "l_point_tc0.hlsl"
Loaded pixelshader "l_spot_b0c0.hlsl"
Loaded vertexdecl "ppcc0t0t0nn"
Loaded technique "l_spot_b0c0"
Loaded pixelshader "l_omni_b0c0.hlsl"
Loaded technique "l_omni_b0c0"
Loaded vertexshader "vertcol_untextured.hlsl"
Loaded technique "wireframe_solid"
Loaded techniqueset "sm2/tools"
Loaded techniqueset "tools"
Loaded material "$additive"
Loaded material "$glare_blind"
Loaded material "$point"
Loaded material "$line"
Loaded material "$line_nodepth"
Loaded vertexshader "transform_only.hlsl"
Loaded pixelshader "black.hlsl"
Loaded vertexdecl "pp"
Loaded technique "clear_alpha_stencil"
Loaded techniqueset "sm2/clear_alpha_stencil"
Loaded techniqueset "clear_alpha_stencil"
Loaded material "clear_alpha_stencil"
Loaded pixelshader "null.hlsl"
Loaded technique "zprepass"
Loaded vertexshader "floatz_build.hlsl"
Loaded pixelshader "floatz_build.hlsl"
Loaded technique "build_floatz"
Loaded technique "build_shadowmap_depth"
Loaded vertexshader "shadowmap_build.hlsl"
Loaded pixelshader "shadowmap_build_sm2.hlsl"
Loaded technique "build_shadowmap_color_sm2"
Loaded techniqueset "sm2/depthprepass"
Loaded pixelshader "shadowmap_build_sm3.hlsl"
Loaded technique "build_shadowmap_color_sm3"
Loaded techniqueset "depthprepass"
Loaded material "depthprepass"
Loaded pixelshader "floatz_clear.hlsl"
Loaded technique "build_floatz_clear"
Loaded techniqueset "sm2/shadowclear"
Loaded techniqueset "shadowclear"
Loaded material "shadowclear"
Loaded vertexshader "shadowcookie_display.hlsl"
Loaded pixelshader "shadowcookie_display.hlsl"
Loaded vertexdecl "ppt0t0"
Loaded technique "shadowcookie_display"
Loaded techniqueset "sm2/shadowcookieoverlay"
Loaded techniqueset "shadowcookieoverlay"
Loaded material "shadowcookieoverlay"
Loaded vertexshader "shadowcookie_blur.hlsl"
Loaded pixelshader "shadowcookie_blur.hlsl"
Loaded technique "shadowcookie_blur"
Loaded techniqueset "sm2/shadowcookieblur"
Loaded techniqueset "shadowcookieblur"
Loaded material "shadowcookieblur"
Loaded vertexshader "shell_shock.hlsl"
Loaded pixelshader "shell_shock.hlsl"
Loaded technique "shell_shock"
Loaded techniqueset "sm2/shell_shock"
Loaded techniqueset "shell_shock"
Loaded material "shellshock"
Loaded vertexshader "shell_shock_flashed.hlsl"
Loaded pixelshader "shell_shock_flashed.hlsl"
Loaded technique "shell_shock_flashed"
Loaded techniqueset "sm2/shell_shock_flashed"
Loaded techniqueset "shell_shock_flashed"
Loaded material "shellshock_flashed"
Loaded vertexshader "color_channel_mixer.hlsl"
Loaded pixelshader "color_channel_mixer.hlsl"
Loaded technique "color_channel_mixer"
Loaded techniqueset "sm2/color_channel_mixer"
Loaded techniqueset "color_channel_mixer"
Loaded material "color_channel_mixer"
Loaded material "frame_color_debug"
Loaded material "frame_alpha_debug"
Loaded pixelshader "vertcol_notexalpha.hlsl"
Loaded technique "passthru_vertalpha"
Loaded techniqueset "sm2/passthru_vertalpha"
Loaded techniqueset "passthru_vertalpha"
Loaded material "feedbackblend"
Loaded pixelshader "vertcol_film.hlsl"
Loaded technique "passthru_film"
Loaded techniqueset "sm2/passthru_film"
Loaded techniqueset "passthru_film"
Loaded material "feedbackfilmblend"
Loaded vertexshader "textured_simple.hlsl"
Loaded pixelshader "textured_simple.hlsl"
Loaded technique "passthru_alpha"
Loaded techniqueset "sm2/passthru_alpha"
Loaded techniqueset "passthru_alpha"
Loaded material "feedbackreplace"
Loaded vertexshader "glow_setup.hlsl"
Loaded pixelshader "glow_setup.hlsl"
Loaded technique "glow_setup"
Loaded techniqueset "sm2/glow_setup"
Loaded techniqueset "glow_setup"
Loaded material "glow_setup"
Loaded vertexshader "glow_consistent_setup.hlsl"
Loaded pixelshader "glow_consistent_setup.hlsl"
Loaded technique "glow_consistent_setup"
Loaded techniqueset "sm2/glow_consistent_setup"
Loaded techniqueset "glow_consistent_setup"
Loaded material "glow_consistent_setup"
Loaded vertexshader "glow_apply_bloom.hlsl"
Loaded pixelshader "glow_apply_bloom.hlsl"
Loaded technique "glow_apply_bloom"
Loaded techniqueset "sm2/glow_apply_bloom"
Loaded techniqueset "glow_apply_bloom"
Loaded material "glow_apply_bloom"
Loaded vertexshader "glow_apply_sky_bleed.hlsl"
Loaded pixelshader "glow_apply_sky_bleed.hlsl"
Loaded technique "glow_apply_sky_bleed"
Loaded techniqueset "sm2/glow_apply_sky_bleed"
Loaded techniqueset "glow_apply_sky_bleed"
Loaded material "glow_apply_sky_bleed"
Loaded vertexshader "filter_symmetric_1.hlsl"
Loaded pixelshader "filter_symmetric_1.hlsl"
Loaded technique "filter_symmetric_1"
Loaded techniqueset "sm2/filter_symmetric_1"
Loaded techniqueset "filter_symmetric_1"
Loaded material "filter_symmetric_1"
Loaded vertexshader "filter_symmetric_2.hlsl"
Loaded pixelshader "filter_symmetric_2.hlsl"
Loaded technique "filter_symmetric_2"
Loaded techniqueset "sm2/filter_symmetric_2"
Loaded techniqueset "filter_symmetric_2"
Loaded material "filter_symmetric_2"
Loaded vertexshader "filter_symmetric_3.hlsl"
Loaded pixelshader "filter_symmetric_3.hlsl"
Loaded technique "filter_symmetric_3"
Loaded techniqueset "sm2/filter_symmetric_3"
Loaded techniqueset "filter_symmetric_3"
Loaded material "filter_symmetric_3"
Loaded vertexshader "filter_symmetric_4.hlsl"
Loaded pixelshader "filter_symmetric_4.hlsl"
Loaded technique "filter_symmetric_4"
Loaded techniqueset "sm2/filter_symmetric_4"
Loaded techniqueset "filter_symmetric_4"
Loaded material "filter_symmetric_4"
Loaded vertexshader "filter_symmetric_5.hlsl"
Loaded pixelshader "filter_symmetric_5.hlsl"
Loaded technique "filter_symmetric_5"
Loaded techniqueset "sm2/filter_symmetric_5"
Loaded techniqueset "filter_symmetric_5"
Loaded material "filter_symmetric_5"
Loaded vertexshader "filter_symmetric_6.hlsl"
Loaded pixelshader "filter_symmetric_6.hlsl"
Loaded technique "filter_symmetric_6"
Loaded techniqueset "sm2/filter_symmetric_6"
Loaded techniqueset "filter_symmetric_6"
Loaded material "filter_symmetric_6"
Loaded vertexshader "filter_symmetric_7.hlsl"
Loaded pixelshader "filter_symmetric_7.hlsl"
Loaded technique "filter_symmetric_7"
Loaded techniqueset "sm2/filter_symmetric_7"
Loaded techniqueset "filter_symmetric_7"
Loaded material "filter_symmetric_7"
Loaded vertexshader "filter_symmetric_8.hlsl"
Loaded pixelshader "filter_symmetric_8.hlsl"
Loaded technique "filter_symmetric_8"
Loaded techniqueset "sm2/filter_symmetric_8"
Loaded techniqueset "filter_symmetric_8"
Loaded material "filter_symmetric_8"
Loaded pixelshader "postfx_dof.hlsl"
Loaded technique "postfx_dof"
Loaded techniqueset "sm2/postfx_dof"
Loaded techniqueset "postfx_dof"
Loaded material "postfx_dof"
Loaded pixelshader "postfx_dof_color.hlsl"
Loaded technique "postfx_dof_color"
Loaded techniqueset "sm2/postfx_dof_color"
Loaded techniqueset "postfx_dof_color"
Loaded material "postfx_dof_color"
Loaded pixelshader "postfx_color.hlsl"
Loaded technique "postfx_color"
Loaded techniqueset "sm2/postfx_color"
Loaded techniqueset "postfx_color"
Loaded material "postfx_color"
Loaded pixelshader "postfx.hlsl"
Loaded technique "postfx"
Loaded techniqueset "sm2/postfx"
Loaded techniqueset "postfx"
Loaded material "postfx"
Loaded vertexshader "dof_downsample.hlsl"
Loaded pixelshader "dof_downsample.hlsl"
Loaded technique "dof_downsample"
Loaded techniqueset "sm2/dof_downsample"
Loaded techniqueset "dof_downsample"
Loaded material "dof_downsample"
Loaded pixelshader "dof_near_coc.hlsl"
Loaded technique "dof_near_coc"
Loaded techniqueset "sm2/dof_near_coc"
Loaded techniqueset "dof_near_coc"
Loaded material "dof_near_coc"
Loaded vertexshader "small_blur.hlsl"
Loaded pixelshader "small_blur.hlsl"
Loaded technique "small_blur"
Loaded techniqueset "sm2/small_blur"
Loaded techniqueset "small_blur"
Loaded material "small_blur"
Loaded image "statmon_warning_tris"
Loaded material "statmon_warning_tris"
Loaded image "console"
Loaded material "console"
Loaded image "warning@file"
Loaded material "code_warning_file"
Loaded image "warning@fps"
Loaded material "code_warning_fps"
Loaded image "voice_on_dim_#0"
Loaded material "nottalkingicon"
Loaded image "voice_on_#0"
Loaded material "talkingicon"
Loaded vertexshader "vertcol_simple_atest.hlsl"
Loaded pixelshader "vertcol_simple_atest.hlsl"
Loaded technique "build_shadowmap_depth00"
Loaded vertexshader "shadowmap_build_atest.hlsl"
Loaded pixelshader "shadowmap_build_atest_sm2.hlsl"
Loaded technique "build_shadowmap_color0_sm2"
Loaded vertexshader "vertcol_simple_fog.hlsl"
Loaded pixelshader "vertcol_simple_fog.hlsl"
Loaded technique "vertcol_simple_fog"
Loaded techniqueset "sm2/shadowcaster"
Loaded techniqueset "shadowcaster"
Loaded image "shadow"
Loaded material "shadowcaster"
Loaded pixelshader "shadowmap_display.hlsl"
Loaded technique "shadowoverlay"
Loaded techniqueset "sm2/shadowoverlay"
Loaded techniqueset "shadowoverlay"
Loaded material "shadowoverlay"
Loaded vertexshader "floatz_display.hlsl"
Loaded pixelshader "floatz_display.hlsl"
Loaded technique "floatzdisplay"
Loaded techniqueset "sm2/floatzdisplay"
Loaded techniqueset "floatzdisplay"
Loaded material "floatz_display"
Loaded vertexshader "floatz.hlsl"
Loaded pixelshader "floatz.hlsl"
Loaded technique "floatz"
Loaded techniqueset "sm2/floatz"
Loaded techniqueset "floatz"
Loaded material "$floatz"
Loaded vertexshader "processed_floatz.hlsl"
Loaded pixelshader "processed_floatz.hlsl"
Loaded technique "processed_floatz"
Loaded techniqueset "sm2/processed_floatz"
Loaded techniqueset "processed_floatz"
Loaded material "$processed_floatz"
Loaded vertexshader "transform_only_tools.hlsl"
Loaded technique "stencilshadow"
Loaded techniqueset "sm2/stencilshadow"
Loaded techniqueset "stencilshadow"
Loaded material "stencilshadow"
Loaded technique "stencildisplay"
Loaded techniqueset "sm2/stencildisplay"
Loaded techniqueset "stencildisplay"
Loaded material "stencildisplay"
Loaded vertexshader "pixel_cost_color_code.hlsl"
Loaded pixelshader "pixel_cost_color_code.hlsl"
Loaded technique "pixel_cost_color_code"
Loaded techniqueset "sm2/pixel_cost_color_code"
Loaded techniqueset "pixel_cost_color_code"
Loaded material "pixel_cost_color_code"
Loaded pixelshader "pixel_cost_add.hlsl"
Loaded technique "pixel_cost_add"
Loaded techniqueset "sm2/pixel_cost_add"
Loaded techniqueset "pixel_cost_add"
Loaded material "pixel_cost_add_depth_always"
Loaded material "pixel_cost_add_depth_disable"
Loaded material "pixel_cost_add_depth_equal"
Loaded material "pixel_cost_add_depth_less"
Loaded material "pixel_cost_add_depthwrite"
Loaded material "pixel_cost_add_nodepthwrite"
Loaded image "line_horizontal"
Loaded material "line_horizontal"
Loaded vertexshader "cinematic.hlsl"
Loaded pixelshader "cinematic.hlsl"
Loaded technique "cinematic"
Loaded techniqueset "sm2/cinematic"
Loaded techniqueset "cinematic"
Loaded image "popups_alpha"
Loaded material "popmenu_bg"
Loaded menu "error_popmenu"
Loaded image "button_highlight_end"
Loaded material "button_highlight_end"
Loaded image "gradient_fadein"
Loaded material "gradient_fadein"
Loaded menu "save_corrupted"
Loaded menu "popup_gettingdata"
Loaded menu "error_popmenu_party"
Loaded menu "error_popmenu_lobby"
Loaded menu "popup_reconnectingtoparty"
Loaded menu "acceptinvite_warning"
Loaded material "cinematic"
Loaded menu "pregame_loaderror"
Loaded menu "pregame_loaderror_mp"
Loaded menu "briefing"
Loaded menu "victoryscreen"
Loaded menu "victory_notice_popmenu"
Loaded menu "connect"
Loaded menulist "ui/code.txt"
Loaded menu "default_menu"
Loaded menulist "ui/default.menu"
Loaded weapon "none"
Loaded xanim "player_mantle_up_57"
Loaded xanim "player_mantle_up_51"
Loaded xanim "player_mantle_up_45"
Loaded xanim "player_mantle_up_39"
Loaded xanim "player_mantle_up_33"
Loaded xanim "player_mantle_up_27"
Loaded xanim "player_mantle_up_21"
Loaded xanim "player_mantle_over_high"
Loaded xanim "player_mantle_over_low"
Loaded xanim "player_mantle_over_mid"
Loaded xanim "void"
Loaded techniqueset "sm2/mc_effect_nofog"
Loaded techniqueset "mc_effect_nofog"
Loaded material "mc/gfx_missing_fx"
ERROR: Failed to load GLTF: Skins must have exactly one common root node
ERROR: Cannot load xmodel "void": Failure while trying to load model for lod 0: "model_export/void_lod0.glb"
ERROR: Failed to load xmodel "void"
ERROR: Could not load asset "void" of type "xmodel"
Unloaded zone "code_post_gfx"
Failed with 0 warnings, 4 errors
```

</details>
OpenAssetTools latest, build from source code. If i remove exception, archive created but game has black screen. Maybe problem here is much deeper than inability to convert model.

Build commad 
```
@echo off
setlocal
set "GAME_FOLDER=D:\Games\Call of Duty 4 Modern Warfare"
set "PROJECT_NAME=code_post_gfx"
set "LOAD_ZONE=%GAME_FOLDER%\zone\english\code_post_gfx.ff"
set "BIN_DIR=%~dp0"
set "BIN_DIR=%BIN_DIR:~0,-1%"

set "LINKER_EXE=%BIN_DIR%\Linker.exe"

set "PROJECT_DIR=D:\Projects\Windows\OpenAssetTools\build\bin\Debug_x64\zone_dump\zone_raw\code_post_gfx"

if not exist "%LINKER_EXE%" (
    echo [ERROR] Linker.exe not found at: "%LINKER_EXE%"
    pause
    exit /b
)

echo [INFO] Bin folder: "%BIN_DIR%"
echo [INFO] Project folder: "%PROJECT_DIR%"
echo [INFO] Starting Linker for project: %PROJECT_NAME%

"%LINKER_EXE%" --verbose ^
  --load "%LOAD_ZONE%" ^
  --base-folder "%PROJECT_DIR%" ^
  --source-search-path "%PROJECT_DIR%\zone_source" ^
  --asset-search-path "%PROJECT_DIR%" ^
  %PROJECT_NAME%

if %ERRORLEVEL% NEQ 0 (
    echo.
    echo [ERROR] Build failed with error code %ERRORLEVEL%
    pause
) else (
    echo.
    echo [SUCCESS] Build finished successfully!
    explorer "%PROJECT_DIR%\zone_out"
)

endlocal
```
Original archive

[code_post_gfx.zip](https://github.com/user-attachments/files/26577251/code_post_gfx.zip)

--- a/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
+++ b/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
@@ -405,13 +405,12 @@ namespace
             return true;
         }
 
-        static std::optional<unsigned> GetRootNodeForSkin(const JsonRoot& jRoot, const JsonSkin& skin)
+        static std::vector<unsigned> GetRootNodesForSkin(const JsonRoot& jRoot, const JsonSkin& skin)
         {
             if (!jRoot.nodes || skin.joints.empty())
-                return std::nullopt;
+                return {};
 
             const auto jointCount = skin.joints.size();
-            auto rootCount = jointCount;
             std::vector<bool> isRoot(jointCount, true);
 
             for (const auto joint : skin.joints)
@@ -431,23 +430,20 @@ namespace
                             if (isRoot[foundChildJointIndex])
                             {
                                 isRoot[foundChildJointIndex] = false;
-                                rootCount--;
                             }
                         }
                     }
                 }
             }
 
-            if (rootCount != 1)
-                throw GltfLoadException("Skins must have exactly one common root node");
-
+            std::vector<unsigned> result;
             for (auto index = 0u; index < jointCount; index++)
             {
                 if (isRoot[index])
-                    return skin.joints[index];
+                    result.emplace_back(skin.joints[index]);
             }
 
-            return std::nullopt;
+            return std::move(result);
         }
 
         void ApplyNodeMatrixTRS(const JsonNode& node, float (&localOffsetRhc)[3], float (&localRotationRhc)[4], float (&scaleRhc)[3]) const
@@ -611,7 +607,9 @@ namespace
                 {
                     if (!ConvertJoint(
                             jRoot, skin, common, skinBoneOffset, childIndex, commonBoneOffset, globalTranslationEigenRhc, globalRotationEigenRhc, bone.scale))
+                    {
                         return false;
+                    }
                 }
             }
 
@@ -625,7 +623,10 @@ namespace
             if (!jRoot.nodes)
                 return false;
 
-            const auto rootNode = GetRootNodeForSkin(jRoot, skin).value_or(skin.joints[0]);
+            auto rootNodes = GetRootNodesForSkin(jRoot, skin);
+            if (rootNodes.empty())
+                rootNodes.emplace_back(skin.joints[0]);
+
             const auto skinBoneOffset = static_cast<unsigned>(common.m_bones.size());
             common.m_bones.resize(skinBoneOffset + skin.joints.size());
 
@@ -633,10 +634,16 @@ namespace
             const Eigen::Quaternionf defaultRotation(1.0f, 0.0f, 0.0f, 0.0f);
             constexpr float defaultScale[3]{1.0f, 1.0f, 1.0f};
 
-            if (!ConvertJoint(jRoot, skin, common, skinBoneOffset, rootNode, std::nullopt, defaultTranslation, defaultRotation, defaultScale))
-                return false;
+            for (const auto rootNode : rootNodes)
+            {
+                if (!ConvertJoint(jRoot, skin, common, skinBoneOffset, rootNode, std::nullopt, defaultTranslation, defaultRotation, defaultScale))
+                    return false;
+            }
 
             common.CalculateBoneLocalsFromGlobals();
+
+            // TODO: Reorder bones if necessary and prepare lookup
+
             return true;
         }
 

--- a/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
+++ b/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
@@ -15,6 +15,7 @@
 #include <limits>
 #include <nlohmann/json.hpp>
 #include <numbers>
+#include <numeric>
 #include <string>
 
 using namespace gltf;
@@ -315,7 +316,7 @@ namespace
                     || !colorAccessor->GetFloatVec4(vertexIndex, vertex.color) || !uvAccessor->GetFloatVec2(vertexIndex, vertex.uv)
                     || !jointsAccessor->GetUnsignedVec4(vertexIndex, joints) || !weightsAccessor->GetFloatVec4(vertexIndex, weights))
                 {
-                    return false;
+                    throw GltfLoadException("Failed to load vertex data from accessors");
                 }
 
                 RhcToLhcCoordinates(vertex.coordinates);
@@ -329,7 +330,16 @@ namespace
                     if (std::abs(weights[i]) < std::numeric_limits<float>::epsilon())
                         continue;
 
-                    common.m_bone_weight_data.weights.emplace_back(joints[i], weights[i]);
+                    assert(joints[i] < m_gltf_to_common_joint_index_lookup.size());
+                    if (joints[i] >= m_gltf_to_common_joint_index_lookup.size())
+                    {
+                        throw GltfLoadException(std::format(
+                            "Vertex weight referenced joint {} (there are only {} joints in skin)", joints[i], m_gltf_to_common_joint_index_lookup.size()));
+                    }
+
+                    const auto xmodelBoneIndex = m_gltf_to_common_joint_index_lookup[joints[i]];
+
+                    common.m_bone_weight_data.weights.emplace_back(xmodelBoneIndex, weights[i]);
                     vertexWeights.weightCount++;
                 }
 
@@ -616,6 +626,54 @@ namespace
             return true;
         }
 
+        void ReorderBonesForXModels(XModelCommon& common)
+        {
+            const auto boneCount = common.m_bones.size();
+            m_gltf_to_common_joint_index_lookup.resize(boneCount);
+            auto reorderedBoneIndex = 0u;
+
+            std::deque<unsigned> parentIndicesToTraverse;
+            for (auto boneIndex = 0u; boneIndex < boneCount; ++boneIndex)
+            {
+                const auto& bone = common.m_bones[boneIndex];
+                if (!bone.parentIndex.has_value())
+                {
+                    parentIndicesToTraverse.emplace_back(boneIndex);
+                    m_gltf_to_common_joint_index_lookup[boneIndex] = reorderedBoneIndex++;
+                }
+            }
+
+            while (!parentIndicesToTraverse.empty())
+            {
+                const auto parentIndex = parentIndicesToTraverse.front();
+                parentIndicesToTraverse.pop_front();
+
+                for (auto boneIndex = 0u; boneIndex < boneCount; ++boneIndex)
+                {
+                    const auto& bone = common.m_bones[boneIndex];
+                    if (bone.parentIndex.has_value() && *bone.parentIndex == parentIndex)
+                    {
+                        parentIndicesToTraverse.emplace_back(boneIndex);
+                        m_gltf_to_common_joint_index_lookup[boneIndex] = reorderedBoneIndex++;
+                    }
+                }
+            }
+
+            assert(reorderedBoneIndex == boneCount);
+
+            std::vector<XModelBone> reorderedBones(boneCount);
+            for (size_t boneIndex = 0; boneIndex < boneCount; ++boneIndex)
+            {
+                auto& reorderedBone = reorderedBones[m_gltf_to_common_joint_index_lookup[boneIndex]];
+                reorderedBone = std::move(common.m_bones[boneIndex]);
+
+                if (reorderedBone.parentIndex.has_value())
+                    reorderedBone.parentIndex = m_gltf_to_common_joint_index_lookup[*reorderedBone.parentIndex];
+            }
+
+            common.m_bones = std::move(reorderedBones);
+        }
+
         bool ConvertSkin(const JsonRoot& jRoot, const JsonSkin& skin, XModelCommon& common)
         {
             if (skin.joints.empty())
@@ -627,8 +685,9 @@ namespace
             if (rootNodes.empty())
                 rootNodes.emplace_back(skin.joints[0]);
 
-            const auto skinBoneOffset = static_cast<unsigned>(common.m_bones.size());
-            common.m_bones.resize(skinBoneOffset + skin.joints.size());
+            // Only one skin per GLTF allowed, more would require more complex mapping and reordering
+            assert(common.m_bones.empty());
+            common.m_bones.resize(skin.joints.size());
 
             constexpr Eigen::Vector3f defaultTranslation(0.0f, 0.0f, 0.0f);
             const Eigen::Quaternionf defaultRotation(1.0f, 0.0f, 0.0f, 0.0f);
@@ -636,13 +695,12 @@ namespace
 
             for (const auto rootNode : rootNodes)
             {
-                if (!ConvertJoint(jRoot, skin, common, skinBoneOffset, rootNode, std::nullopt, defaultTranslation, defaultRotation, defaultScale))
+                if (!ConvertJoint(jRoot, skin, common, 0, rootNode, std::nullopt, defaultTranslation, defaultRotation, defaultScale))
                     return false;
             }
 
+            ReorderBonesForXModels(common);
             common.CalculateBoneLocalsFromGlobals();
-
-            // TODO: Reorder bones if necessary and prepare lookup
 
             return true;
         }
@@ -838,7 +896,15 @@ namespace
         std::vector<std::unique_ptr<BufferView>> m_buffer_views;
         std::vector<std::unique_ptr<Buffer>> m_buffers;
 
+        // Old gltf support for OAT used bad formulas to calculate right-handed coordinate system rotations
+        // To make the fixed code be backwards compatible, old behaviour can be restored with this setting.
         bool m_bad_rotation_formulas;
+
+        // We may need to reorder bones to account for the constraints of xmodels:
+        // Root bones have the lowest indices and the index of the parent of each bone must be lower than its own.
+        // The index in this vector is the joint index of the gltf.
+        // The value is the index in the common xmodel.
+        std::vector<unsigned> m_gltf_to_common_joint_index_lookup;
     };
 } // namespace
 

--- a/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
+++ b/src/ObjLoading/XModel/Gltf/GltfLoader.cpp
@@ -12,7 +12,6 @@
 #include <deque>
 #include <exception>
 #include <format>
-#include <iostream>
 #include <limits>
 #include <nlohmann/json.hpp>
 #include <numbers>
@@ -451,7 +450,7 @@ namespace
             return std::nullopt;
         }
 
-        void ApplyNodeMatrixTRS(const JsonNode& node, float (&localOffsetRhc)[3], float (&localRotationRhc)[4], float (&scaleRhc)[3])
+        void ApplyNodeMatrixTRS(const JsonNode& node, float (&localOffsetRhc)[3], float (&localRotationRhc)[4], float (&scaleRhc)[3]) const
         {
             const auto matrix = Eigen::Matrix4f({
                 {(*node.matrix)[0], (*node.matrix)[4], (*node.matrix)[8],  (*node.matrix)[12]},
@@ -492,7 +491,7 @@ namespace
             scaleRhc[2] = matrix.block<3, 1>(0, 2).norm();
         }
 
-        void ApplyNodeSeparateTRS(const JsonNode& node, float (&localOffsetRhc)[3], float (&localRotationRhc)[4], float (&scaleRhc)[3])
+        void ApplyNodeSeparateTRS(const JsonNode& node, float (&localOffsetRhc)[3], float (&localRotationRhc)[4], float (&scaleRhc)[3]) const
         {
             if (node.translation)
             {
@@ -630,7 +629,7 @@ namespace
             const auto skinBoneOffset = static_cast<unsigned>(common.m_bones.size());
             common.m_bones.resize(skinBoneOffset + skin.joints.size());
 
-            const Eigen::Vector3f defaultTranslation(0.0f, 0.0f, 0.0f);
+            constexpr Eigen::Vector3f defaultTranslation(0.0f, 0.0f, 0.0f);
             const Eigen::Quaternionf defaultRotation(1.0f, 0.0f, 0.0f, 0.0f);
             constexpr float defaultScale[3]{1.0f, 1.0f, 1.0f};
 

--- a/src/ObjLoading/XModel/LoaderXModel.cpp.template
+++ b/src/ObjLoading/XModel/LoaderXModel.cpp.template
@@ -323,8 +323,12 @@ namespace
             {
                 xmodel.parentList = m_memory.Alloc<unsigned char>(xmodel.numBones - xmodel.numRootBones);
 
-                // For some reason Treyarch games allocate for a vec4 here. it is treated as a vec3 though?
+#if defined(FEATURE_IW3) || defined(FEATURE_T5) || defined(FEATURE_T6)
+                // For some reason some games allocate for a vec4 here. it is treated as a vec3 though?
                 xmodel.trans = m_memory.Alloc<float>((xmodel.numBones - xmodel.numRootBones) * 4u);
+#else
+                xmodel.trans = m_memory.Alloc<float>((xmodel.numBones - xmodel.numRootBones) * 3u);
+#endif
                 xmodel.quats = m_memory.Alloc<XModelQuat>(xmodel.numBones - xmodel.numRootBones);
             }
             else
@@ -373,7 +377,11 @@ namespace
             // Viewhands seem to have nulled trans for some reason?
             if (jXModel.type.value_or(JsonXModelType::RIGID) == JsonXModelType::VIEWHANDS)
             {
+#if defined(FEATURE_IW3) || defined(FEATURE_T5) || defined(FEATURE_T6)
                 memset(xmodel.trans, 0, sizeof(float) * 4 * (xmodel.numBones - xmodel.numRootBones));
+#else
+                memset(xmodel.trans, 0, sizeof(float) * 3 * (xmodel.numBones - xmodel.numRootBones));
+#endif
             }
 
             return true;

--- a/src/ObjWriting/XModel/Gltf/GltfWriter.cpp
+++ b/src/ObjWriting/XModel/Gltf/GltfWriter.cpp
@@ -164,10 +164,10 @@ namespace
 
             const auto meshCount = xmodel.m_objects.size();
             for (auto meshIndex = 0u; meshIndex < meshCount; meshIndex++)
-                rootNode.children->push_back(m_first_mesh_node + meshIndex);
+                rootNode.children->emplace_back(m_first_mesh_node + meshIndex);
 
-            if (!xmodel.m_bones.empty())
-                rootNode.children->push_back(m_first_bone_node);
+            for (auto rootBoneIndex = 0u; rootBoneIndex < m_root_bone_count; rootBoneIndex++)
+                rootNode.children->emplace_back(m_first_bone_node + rootBoneIndex);
 
             m_root_node = static_cast<unsigned>(gltf.nodes->size());
             gltf.nodes->emplace_back(std::move(rootNode));
@@ -278,6 +278,7 @@ namespace
 
             const auto boneCount = common.m_bones.size();
             m_first_bone_node = static_cast<unsigned>(gltf.nodes->size());
+            m_root_bone_count = 0;
             for (auto boneIndex = 0u; boneIndex < boneCount; boneIndex++)
             {
                 JsonNode boneNode;
@@ -309,6 +310,11 @@ namespace
                     translation -= parentTranslation;
                     translation = inverseParentRotation * translation;
                     rotation = inverseParentRotation * rotation;
+                }
+                else
+                {
+                    assert(m_root_bone_count == boneIndex);
+                    m_root_bone_count++;
                 }
                 rotation.normalize();
 
@@ -679,6 +685,7 @@ namespace
         unsigned m_first_mesh_node = 0u;
         unsigned m_root_node = 0u;
         unsigned m_first_bone_node = 0u;
+        unsigned m_root_bone_count = 0u;
         unsigned m_position_accessor = 0u;
         unsigned m_normal_accessor = 0u;
         unsigned m_uv_accessor = 0u;

--- a/src/ObjWriting/XModel/Gltf/GltfWriter.cpp
+++ b/src/ObjWriting/XModel/Gltf/GltfWriter.cpp
@@ -187,10 +187,10 @@ namespace
 
             const auto meshCount = xmodel.m_objects.size();
             for (auto meshIndex = 0u; meshIndex < meshCount; meshIndex++)
-                rootNode.children->push_back(m_first_mesh_node + meshIndex);
+                rootNode.children->emplace_back(m_first_mesh_node + meshIndex);
 
-            if (!xmodel.m_bones.empty())
-                rootNode.children->push_back(m_first_bone_node);
+            for (auto rootBoneIndex = 0u; rootBoneIndex < m_root_bone_count; rootBoneIndex++)
+                rootNode.children->emplace_back(m_first_bone_node + rootBoneIndex);
 
             m_root_node = static_cast<unsigned>(gltf.nodes->size());
             gltf.nodes->emplace_back(std::move(rootNode));
@@ -303,6 +303,7 @@ namespace
 
             const auto boneCount = common.m_bones.size();
             m_first_bone_node = static_cast<unsigned>(gltf.nodes->size());
+            m_root_bone_count = 0;
             for (auto boneIndex = 0u; boneIndex < boneCount; boneIndex++)
             {
                 JsonNode boneNode;
@@ -334,6 +335,11 @@ namespace
                     translation -= parentTranslation;
                     translation = inverseParentRotation * translation;
                     rotation = inverseParentRotation * rotation;
+                }
+                else
+                {
+                    assert(m_root_bone_count == boneIndex);
+                    m_root_bone_count++;
                 }
                 rotation.normalize();
 
@@ -745,6 +751,7 @@ namespace
         unsigned m_first_mesh_node = 0u;
         unsigned m_root_node = 0u;
         unsigned m_first_bone_node = 0u;
+        unsigned m_root_bone_count = 0u;
         unsigned m_position_accessor = 0u;
         unsigned m_normal_accessor = 0u;
         unsigned m_color_accessor = 0u;

